File: proj/compiler/test_programs/local_laplacian.py
import numpy as np
#import skimage
#import skimage.io
#import skimage.color
import sys; sys.path += ['../../compiler']
import util

#transform(TypeSpecialize(checks=False))
def local_laplacian(input_img):

    levels = 8
    alpha = 1.0 / 7.0
    beta = 1
    J = 8
    eps = 0.01
    
    gray = 0.299 * input_img[:, :, 0] + 0.587 * input_img[:, :, 1] + 0.114 * input_img[:, :, 2]
    
    w = input_img.shape[0]
    h = input_img.shape[1]
    
    gPyramid = np.zeros([w, h, levels, J])
    lPyramid = np.zeros([w, h, levels, J])
    inGPyramid = np.zeros([w, h, J])
    outLPyramid = np.zeros([w, h, J])
    outGPyramid = np.zeros([w, h, J])
    
    output_img = np.zeros([w, h, 3])
    
    downx = np.zeros([w, h, levels])
    gdownx = np.zeros([w, h])
    upx = np.zeros([w, h, levels])
    upy = np.zeros([w, h, levels])
    gupx = np.zeros([w, h])
    gupy = np.zeros([w, h])
    
#transform(Parallel())
    for k in range(levels):
        for x in range(w):
            for y in range(h):
            
                if x == 2000 and y == 800:
                    x = 2000
                level = k * (1.0 / (levels - 1))
                idx = int(gray[x, y] * (levels - 1) * 256.0)
                idx = np.clip(idx, 0, 256 * (levels - 1))
                fx = (idx - 256.0 * k) / 256.0
                gPyramid[x, y, k, 0] = beta * (gray[x, y] - level) + level + alpha * fx * np.exp(-fx * fx / 2.0)
    
    for j in range(1, J):
    
        w_j = int(w * 2 ** (-j))
        h_j = int(h * 2 ** (-j))
        
#transform(Parallel())
        for x in range(w_j * 2):
            for y in range(1, h_j - 1):
                
                downx[x, y, :] = (gPyramid[x, 2 * y - 1, :, j - 1] + 3.0 * (gPyramid[x, 2 * y, :, j - 1] + gPyramid[x, 2 * y + 1, :, j - 1]) + gPyramid[x, 2 * y + 2, :, j - 1]) / 8.0
            
            downx[x, 0, :] = (4.0 * gPyramid[x, 0, :, j - 1] + 3.0 * gPyramid[x, 1, :, j - 1] + gPyramid[x, 2, :, j - 1]) / 8.0
            downx[x, h_j - 1, :] = (gPyramid[x, h_j * 2 - 3, :, j - 1] + 3.0 * gPyramid[x, h_j * 2 - 2, :, j - 1] + 4.0 * gPyramid[x, h_j * 2 - 1, :, j - 1]) / 8.0
        
#transform(Parallel())
        for y in range(h_j):
            for x in range(1, w_j - 1):
            
                gPyramid[x, y, :, j] = (downx[2 * x - 1, y, :] + 3.0 * (downx[2 * x, y, :] + downx[2 * x + 1, y, :]) + downx[2 * x + 2, y, :]) / 8.0
            
            gPyramid[0, y, :, j] = (4.0 * downx[0, y, :] + 3.0 * downx[1, y, :] + downx[2, y, :]) / 8.0
            gPyramid[w_j - 1, y, :, j] = (downx[w_j * 2 - 3, y, :] + 3.0 * downx[w_j * 2 - 2, y, :] + 4.0 * downx[w_j * 2 - 1, y, :]) / 8.0
    
    lPyramid[:, :, :, J - 1] = gPyramid[:, :, :, J - 1]
    
    for j in range(J - 2, -1, -1):
        
        w_j = int(w * 2 ** (-j))
        h_j = int(h * 2 ** (-j))
        
#transform(Parallel())
        for x in range(int(w_j / 2)):
            for y in range(1, h_j):
            
                upx[x, y, :] = 0.25 * gPyramid[x, int((y + 1) / 2) - 1, :, j + 1] + 0.75 * gPyramid[x, int(y / 2), :, j + 1]
            
            upx[x, 0, :] = gPyramid[x, 0, :, j + 1]
        
#transform(Parallel())
        for y in range(h_j):
            for x in range(1, w_j):
            
                upy[x, y, :] = 0.25 * upx[int((x + 1) / 2) - 1, y, :] + 0.75 * upx[int(x / 2), y, :]
                lPyramid[x, y, :, j] = gPyramid[x, y, :, j] - upy[x, y, :]
            
            upy[0, y, :] = upx[0, y, :]
            lPyramid[0, y, :, j] = gPyramid[0, y, :, j] - upy[0, y, :]
    
    inGPyramid[:, :, 0] = gray
    
    for j in range(1, J):
    
        w_j = int(w * 2 ** (-j))
        h_j = int(h * 2 ** (-j))
        
#transform(Parallel())
        for x in range(w_j * 2):
            for y in range(1, h_j - 1):
            
                gdownx[x, y] = (inGPyramid[x, 2 * y - 1, j - 1] + 3.0 * (inGPyramid[x, 2 * y, j - 1] + inGPyramid[x, 2 * y + 1, j - 1]) + inGPyramid[x, 2 * y + 2, j - 1]) / 8.0
            
            gdownx[x, 0] = (4.0 * inGPyramid[x, 0, j - 1] + 3.0 * inGPyramid[x, 1, j - 1] + inGPyramid[x, 2, j - 1]) / 8.0
            gdownx[x, h_j - 1] = (inGPyramid[x, h_j * 2 - 3, j - 1] + 3.0 * inGPyramid[x, h_j * 2 - 2, j - 1] + 4.0 * inGPyramid[x, h_j * 2 - 1, j - 1]) / 8.0
        
#transform(Parallel())
        for y in range(h_j):
            for x in range(1, w_j - 1):
            
                inGPyramid[x, y, j] = (gdownx[2 * x - 1, y] + 3.0 * (gdownx[2 * x, y] + gdownx[2 * x + 1, y]) + gdownx[2 * x + 2, y]) / 8.0
            
            inGPyramid[0, y, j] = (4.0 * gdownx[0, y] + 3.0 * gdownx[1, y] + gdownx[2, y]) / 8.0
            inGPyramid[w_j - 1, y, j] = (gdownx[w_j * 2 - 3, y] + 3.0 * gdownx[w_j * 2 - 2, y] + 4.0 * gdownx[w_j * 2 - 1, y]) / 8.0
    
    for j in range(J):
        
        w_j = int(w * 2 ** (-j))
        h_j = int(h * 2 ** (-j))
        
#transform(Parallel())
        for x in range(w_j):
            for y in range(h_j):
            
                level = inGPyramid[x, y, j] * (levels - 1.0)
                li = np.clip(int(level), 0, levels - 2)
                lf = level - li
                outLPyramid[x, y, j] = (1.0 - lf) * lPyramid[x, y, li, j] + lf * lPyramid[x, y, li + 1, j]
    
    outGPyramid[:, :, J - 1] = outLPyramid[:, :, J - 1]
    
    for j in range(J - 2, -1, -1):
    
        w_j = int(w * 2 ** (-j))
        h_j = int(h * 2 ** (-j))
        
#transform(Parallel())
        for x in range(int(w_j / 2)):
            for y in range(1, h_j):
            
                gupx[x, y] = 0.25 * outGPyramid[x, int((y + 1) / 2) - 1, j + 1] + 0.75 * outGPyramid[x, int(y / 2), j + 1]
    
            gupx[x, 0] = outGPyramid[x, 0, j + 1]
            
#transform(Parallel())
        for y in range(h_j):
            for x in range(1, w_j):
                
                gupy[x, y] = 0.25 * gupx[int((x + 1) / 2) - 1, y] + 0.75 * gupx[int(x / 2), y]
                outGPyramid[x, y, j] = gupy[x, y] + outLPyramid[x, y, j]
                    
            gupy[0, y] = gupx[0, y]
            outGPyramid[0, y, j] = gupy[0, y] + outLPyramid[0, y, j]
    
#transform(Parallel())
    for x in range(w):
        for y in range(h):
            
            output_img[x, y, :] = outGPyramid[x, y, 0] * (input_img[x, y, :] + eps) / (gray[x, y] + eps)
    
    output_img = np.clip(output_img, 0, 1)
    
    return output_img

input_img = util.image_filename('small_temple_rgb.png')
#input = skimage.io.imread(input_img)
#input = skimage.img_as_float(input)

#output = local_laplacian(input)

#skimage.io.imsave('result.png', np.clip(output, 0, 1))

def test(n = None):
    ans = util.test_image_pipeline_filename(local_laplacian, (input_img,), n, name = 'local_laplacian')
    return util.combine_tests([ans])

if __name__ == '__main__':
    test()

File: proj/apps/local_laplacian/c/local_laplacian.c
#include "c_util.h"
#include <stdio.h>
#include <omp.h>

float max(float a, float b)
{
    return a > b ? a : b;
}

float min(float a, float b)
{
    return a < b ? a : b;
}

float clip(float input, float lower_bound, float upper_bound)
{
    return min(max(input, lower_bound), upper_bound);
}

void local_laplacian(FLOAT3_IMG *input_img, FLOAT3_IMG *output_img)
{
    int levels = 8;
    float alpha = 1.0f / 7.0f;
    float beta = 1.0f;
    int J = 8;
    float eps = 0.01f;
    
    int r, c;
    int i, j;
    
    FLOAT_IMG *gray = zeros_gray(input_img->rows, input_img->cols);
    
    #pragma omp parallel for private(c)
    for(r = 0; r < input_img->rows; r++)
    {
        for(c = 0; c < input_img->cols; c++)
        {
            gray->data[r][c] = 0.299f * input_img->data[r][c][0] + 0.587f * input_img->data[r][c][1] + 0.114f * input_img->data[r][c][2];
        }
    }
    
    int w = input_img->rows;
    int h = input_img->cols;
    
    real ****gPyramid;
    real ****lPyramid;
    real ***inGPyramid;
    real ***outLPyramid;
    real ***outGPyramid;
    
    real *gPyramid_data = malloc(input_img->rows * input_img->cols * levels * J * sizeof(real));
    real **gPyramid_pointer = malloc(input_img->rows * sizeof(real ***) + input_img->rows * input_img->cols * sizeof(real **) + input_img->rows * input_img->cols * levels * sizeof(real *));
    gPyramid = (real ****)gPyramid_pointer;
    
    real *lPyramid_data = malloc(input_img->rows * input_img->cols * levels * J * sizeof(real));
    real **lPyramid_pointer = malloc(input_img->rows * sizeof(real ***) + input_img->rows * input_img->cols * sizeof(real **) + input_img->rows * input_img->cols * levels * sizeof(real *));
    lPyramid = (real ****)lPyramid_pointer;
    
    real *inGPyramid_data = malloc(input_img->rows * input_img->cols * J * sizeof(real));
    real **inGPyramid_pointer = malloc(input_img->rows * sizeof(real **) + input_img->rows * input_img->cols * sizeof(real *));
    inGPyramid = (real ***)inGPyramid_pointer;
    
    real *outLPyramid_data = malloc(input_img->rows * input_img->cols * J * sizeof(real));
    real **outLPyramid_pointer = malloc(input_img->rows * sizeof(real **) + input_img->rows * input_img->cols * sizeof(real *));
    outLPyramid = (real ***)outLPyramid_pointer;
    
    real *outGPyramid_data = malloc(input_img->rows * input_img->cols * J * sizeof(real));
    real **outGPyramid_pointer = malloc(input_img->rows * sizeof(real **) + input_img->rows * input_img->cols * sizeof(real *));
    outGPyramid = (real ***)outGPyramid_pointer;
    
    #pragma omp parallel for private(c, i, j)
    for(r = 0; r < input_img->rows; r++)
    {
        gPyramid[r] = (real ***)&gPyramid_pointer[input_img->rows + r * input_img->cols];
        lPyramid[r] = (real ***)&lPyramid_pointer[input_img->rows + r * input_img->cols];
        inGPyramid[r] = (real **)&inGPyramid_pointer[input_img->rows + r * input_img->cols];
        outLPyramid[r] = (real **)&outLPyramid_pointer[input_img->rows + r * input_img->cols];
        outGPyramid[r] = (real **)&outGPyramid_pointer[input_img->rows + r * input_img->cols];
    
        for(c = 0; c < input_img->cols; c++)
        {
            gPyramid[r][c] = (real **)&gPyramid_pointer[input_img->rows + input_img->rows * input_img->cols + r * input_img->cols * levels + c * levels];
            lPyramid[r][c] = (real **)&lPyramid_pointer[input_img->rows + input_img->rows * input_img->cols + r * input_img->cols * levels + c * levels];
            inGPyramid[r][c] = &inGPyramid_data[J * (r * input_img->cols + c)];
            outLPyramid[r][c] = &outLPyramid_data[J * (r * input_img->cols + c)];
            outGPyramid[r][c] = &outGPyramid_data[J * (r * input_img->cols + c)];
        
            for(i = 0; i < levels; i++)
            {
                gPyramid[r][c][i] = &gPyramid_data[J * (r * input_img->cols * levels + c * levels + i)];
                lPyramid[r][c][i] = &lPyramid_data[J * (r * input_img->cols * levels + c * levels + i)];
        
                for(j = 0; j < J; j++)
                {
                    gPyramid[r][c][i][j] = 0.0f;
                    lPyramid[r][c][i][j] = 0.0f;
                }
            }
        
            for(j = 0; j < J; j++)
            {
                inGPyramid[r][c][j] = 0.0f;
                outLPyramid[r][c][j] = 0.0f;
                outGPyramid[r][c][j] = 0.0f;
            }
        }
    }
    
    int k;
    float level;
    int idx;
    float fx;
    
    #pragma omp parallel for private(r, c, level, idx, fx)
    for(k = 0; k < levels; k++)
    {
        for(r = 0; r < input_img->rows; r++)
        {
            for(c = 0; c < input_img->cols; c++)
            {
                level = k * (1.0f / (levels - 1.0f));
                idx = gray->data[r][c] * (levels - 1) * 256;
                idx = clip(idx, 0, 256 * (levels - 1));
                fx = (idx - 256.0f * k) / 256.0f;
                gPyramid[r][c][k][0] = beta * (gray->data[r][c] - level) + level + alpha * fx * exp(-fx * fx / 2.0f);
            }
        }
    }
    
    #pragma omp parallel for private(c)
    for(r = 0; r < input_img->rows; r++)
    {
        for(c = 0; c < input_img->cols; c++)
        {
            inGPyramid[r][c][0] = gray->data[r][c];
        }
    }
    
    int idx_r1, idx_r2, idx_c1, idx_c2;
    int w_j, h_j;
    
    for(j = 1; j < J; j++)
    {
        w_j = input_img->rows >> j;
        h_j = input_img->cols >> j;
        
        #pragma omp parallel for private(c, i, idx_r1, idx_r2, idx_c1, idx_c2)
        for(r = 0; r < w_j; r++)
        {
            for(c = 1; c < h_j; c++)
            {
                idx_r1 = 2 * r - 1;
                idx_r2 = 2 * r + 2;
                idx_c1 = 2 * c - 1;
                idx_c2 = 2 * c + 2;
                
                if (r == 0)
                    idx_r1 = 0;
                if (r == w_j - 1)
                    idx_r2 = w_j * 2 - 1;
                if (c == 0)
                    idx_c1 = 0;
                if (c == h_j - 1)
                    idx_c2 = h_j * 2 - 1;
                
                for (i = 0; i < levels; i++)
                {
                    gPyramid[r][c][i][j] = (gPyramid[idx_r1][idx_c1][i][j - 1] + 3.0f * gPyramid[idx_r1][2 * c][i][j - 1] + 3.0f * gPyramid[idx_r1][2 * c + 1][i][j - 1] + gPyramid[idx_r1][idx_c2][i][j - 1] + 3.0f * gPyramid[2 * r][idx_c1][i][j - 1] + 9.0f * gPyramid[2 * r][2 * c][i][j - 1] + 9.0f * gPyramid[2 * r][2 * c + 1][i][j - 1] + 3.0f * gPyramid[2 * r][idx_c2][i][j - 1] + 3.0f * gPyramid[2 * r + 1][idx_c1][i][j - 1] + 9.0f * gPyramid[2 * r + 1][2 * c][i][j - 1] + 9.0f * gPyramid[2 * r + 1][2 * c + 1][i][j - 1] + 3.0f * gPyramid[2 * r + 1][idx_c2][i][j - 1] + gPyramid[idx_r2][idx_c1][i][j - 1] + 3.0f * gPyramid[idx_r2][2 * c][i][j - 1] + 3.0f * gPyramid[idx_r2][2 * c + 1][i][j - 1] + gPyramid[idx_r2][idx_c2][i][j - 1]) / 64.0f;
                    if (r == 49 && c == 78 && i == 7 && j == 1)
                        r = 49;
                }
                inGPyramid[r][c][j] = (inGPyramid[idx_r1][idx_c1][j - 1] + 3.0f * inGPyramid[idx_r1][2 * c][j - 1] + 3.0f * inGPyramid[idx_r1][2 * c + 1][j - 1] + inGPyramid[idx_r1][idx_c2][j - 1] + 3.0f * inGPyramid[2 * r][idx_c1][j - 1] + 9.0f * inGPyramid[2 * r][2 * c][j - 1] + 9.0f * inGPyramid[2 * r][2 * c + 1][j - 1] + 3.0f * inGPyramid[2 * r][idx_c2][j - 1] + 3.0f * inGPyramid[2 * r + 1][idx_c1][j - 1] + 9.0f * inGPyramid[2 * r + 1][2 * c][j - 1] + 9.0f * inGPyramid[2 * r + 1][2 * c + 1][j - 1] + 3.0f * inGPyramid[2 * r + 1][idx_c2][j - 1] + inGPyramid[idx_r2][idx_c1][j - 1] + 3.0f * inGPyramid[idx_r2][2 * c][j - 1] + 3.0f * inGPyramid[idx_r2][2 * c + 1][j - 1] + inGPyramid[idx_r2][idx_c2][j - 1]) / 64.0f;
            }
        }
    }
    
    #pragma omp parallel for private(r, c)
    for(r = 0; r < input_img->rows; r++)
    {
        for(c = 0; c < input_img->cols; c++)
        {
            for (k = 0; k < levels; k++)
            {
                lPyramid[r][c][k][J - 1] = gPyramid[r][c][k][J - 1];
            }
        }
    }
    
    j = J - 1;
    w_j = input_img->rows >> j;
    h_j = input_img->cols >> j;
    
    int li;
    float lf;
    
    #pragma omp parallel for private(c, level, li, lf)
    for (r = 0; r < w_j; r++)
    {
        for (c = 0; c < h_j; c++)
        {
            level = inGPyramid[r][c][j] * (levels - 1.0f);
            li = clip((int)level, 0, levels - 2);
            lf = level - li;
            outLPyramid[r][c][j] = (1.0f - lf) * lPyramid[r][c][li][j] + lf * lPyramid[r][c][li + 1][j];
        }
    }
    
    #pragma omp parallel for private(c)
    for(r = 0; r < input_img->rows; r++)
    {
        for(c = 0; c < input_img->cols; c++)
        {
            outGPyramid[r][c][J - 1] = outLPyramid[r][c][J - 1];
        }
    }
    
    for(j = J - 2; j > -1; j--)
    {
        w_j = input_img->rows >> j;
        h_j = input_img->cols >> j;
        
        #pragma omp parallel for private(c, i, idx_r1, idx_c1, level, li, lf)
        for (r = 0; r < w_j; r++)
        {
            for (c = 0; c < h_j; c++)
            {
                idx_r1 = (int)((r + 1) / 2) - 1;
                idx_c1 = (int)((c + 1) / 2) - 1;
                
                if (r == 0)
                    idx_r1 = 0;
                if (c == 0)
                    idx_c1 = 0;
                
                for (i = 0; i < levels; i++)
                {
                    lPyramid[r][c][i][j] = gPyramid[r][c][i][j] - (gPyramid[idx_r1][idx_c1][i][j + 1] + 3.0f * gPyramid[idx_r1][(int)(c / 2)][i][j + 1] + 3.0f * gPyramid[(int)(r / 2)][idx_c1][i][j + 1] + 9.0f * gPyramid[(int)(r / 2)][(int)(c / 2)][i][j + 1]) / 16.0f;
                }
                
                level = inGPyramid[r][c][j] * (levels - 1.0f);
                li = clip((int)level, 0, levels - 2);
                lf = level - li;
                outLPyramid[r][c][j] = (1.0f - lf) * lPyramid[r][c][li][j] + lf * lPyramid[r][c][li + 1][j];
                
                outGPyramid[r][c][j] = outLPyramid[r][c][j] + (outGPyramid[idx_r1][idx_c1][j + 1] + 3.0f * outGPyramid[idx_r1][(int)(c / 2)][j + 1] + 3.0f * outGPyramid[(int)(r / 2)][idx_c1][j + 1] + 9.0f * outGPyramid[(int)(r / 2)][(int)(c / 2)][j + 1]) / 16.0f;
            }
        }
    }
    
    #pragma omp parallel for private(c)
    for(r = 0; r < input_img->rows; r++)
    {
        for(c = 0; c < input_img->cols; c++)
        {
            output_img->data[r][c][0] = clip(outGPyramid[r][c][0] * (input_img->data[r][c][0] + eps) / (gray->data[r][c] + eps), 0, 1);
            output_img->data[r][c][1] = clip(outGPyramid[r][c][0] * (input_img->data[r][c][1] + eps) / (gray->data[r][c] + eps), 0, 1);
            output_img->data[r][c][2] = clip(outGPyramid[r][c][0] * (input_img->data[r][c][2] + eps) / (gray->data[r][c] + eps), 0, 1);
        }
    }
    
    free_img_gray(gray);
    
    free(gPyramid_data);
    free(gPyramid_pointer);
    free(lPyramid_data);
    free(lPyramid_pointer);
    free(inGPyramid_data);
    free(inGPyramid_pointer);
    free(outLPyramid_data);
    free(outLPyramid_pointer);
    free(outGPyramid_data);
    free(outGPyramid_pointer);
}

int main(int argc, char **argv)
{
    FLOAT3_IMG *input_img;
    
    if(argc == 1)
    input_img = read_img_rgb("../../images/small_temple_rgb.png");
    else
    {
    if(argc == 3)
        input_img = read_img_rgb(argv[1]);
    else
    {
        printf("Usage: ./a.out in.png out.png or ./a.out\n");
        return 0;
    }
    }
    
    FLOAT3_IMG *output_img = zeros_rgb(input_img->rows, input_img->cols);
    
    double d1 = wall_time();
    local_laplacian(input_img, output_img);
    double d2 = wall_time();
    
    printf("%f\n", d2 - d1);
    
    if(argc == 1)
    write_img_rgb(output_img, "test.png");
    else
    write_img_rgb(output_img, argv[2]);
    
    free_img_rgb(input_img);
    free_img_rgb(output_img);
    
    return 0;
}